src/blocks/CalligraphyApp.tsx

```diff
@@ -0,0 +1,753 @@
+            colors={COLORS_DEFAULT}
+            onColorSelected={color => props.setActiveColor(color)}
+            activeColor={props.activeColor}
+        />
+      }
+      { props.activeToolbarTab === CalligraphyToolbarView.BACKGROUND
+        && <CalligraphyBackgroundSelector currentBackground={props.currentBackground} setCurrentBackground={props.setCurrentBackground}
+        />
+      }
+      { props.activeToolbarTab === CalligraphyToolbarView.BRUSH
+        && <CalligraphyBrushSelector currentBrush={props.currentBrush} setCurrentBrush={props.setCurrentBrush} />
+      }
+      {}
+    </div>
+  );
+}
+
+// The edit view for the Calligraphy block
+interface CalligraphyEditProps {
+  onSave: () => void;
+  setImageDataURL: (imgDataURL: string) => void
+  width: string;
+}
+const CalligraphyEdit = (props: CalligraphyEditProps) => {
+  const [activeColor, setActiveColor] = useState('#000000');
+  const [canvasClearSwitch, setCanvasClearSwitch] = useState(false);
+  const [canvasUndoSwitch, setCanvasUndoSwitch] = useState(false);
+  const [currentBackground, setCurrentBackground] = useState<CalligraphyBackground>("lines");
+  const [activeToolbarTab, setActivePaletteTab] = useState(CalligraphyToolbarView.COLOR);
+  const [currentBrush, setCurrentBrush] = useState("spray");
+  return (
+    <div className='flex flex-col gap-3 justify-between h-[88vh]'>
+      <CalligraphyCanvas 
+        width={props.width} 
+        canvasClearSwitch={canvasClearSwitch}
+        canvasUndoSwitch={canvasUndoSwitch}
+        activeColor={activeColor} 
+        backgroundStyle={currentBackground}
+        currentBrush={currentBrush}
+        setImageDataURL={props.setImageDataURL}
+      />
+      <div className='flex flex-0 flex-col'>
+        <div className='flex flex-col gap-4'>
+          <CalligraphyToolbarTab
+            activeToolbarTab={activeToolbarTab}
+            activeColor={activeColor}
+            setActiveColor={setActiveColor}
+            currentBackground={currentBackground}
+            setCurrentBackground={setCurrentBackground}
+            currentBrush={currentBrush}
+            setCurrentBrush={setCurrentBrush}
+          />
+          <CalligraphyToolbar
+            activeColor={activeColor}
+            setActiveToolbarTab={setActivePaletteTab}
+            toggleCanvasUndoSwitch={() => setCanvasUndoSwitch(!canvasUndoSwitch)}
+            toggleCanvasClearSwitch={() => setCanvasClearSwitch(!canvasClearSwitch)}
+            activeToolbarTab={activeToolbarTab} />
+        </div>
+        <SeamSaveButton onClick={props.onSave}/>
+      </div>
+    </div>
+  );
+}
+
+// Top level component; wrapper for functional React components
+export default class CalligraphyBlock extends Block {
+
+  render(width?: string, height?: string) {
+    if (Object.keys(this.model.data).length === 0) {
+      return BlockFactory.renderEmptyState(this.model, this.onEditCallback!)
+    }
+
+    const { imageData } = this.model.data;
```

Comment:
it would be better to store the image data at a specified key, like 
`const imageData = this.model.data['imageData'];`
so that you could add other keys later with different information

Reply:
I think the line of code you suggested is isomorphic to the object destructuring assignment we have here unless I'm misunderstanding. Both are accessing the `this.model` object and assigning its `data` property to a newly created `imageData` variable.

Reply:
I believe the object destructuring assignment we have here is isomorphic to your suggestion. Here we are extracting the `imageData` from the model to a local variable for the render function (perhaps redundant since we only use it once in the rendered `<img>`)

Our callback for _writing_ the image data to `this.model.data` (line 741) is ` this.model.data["imageData"] = imageDataURL` creates that key on `model.data`